cocos/base/CCGLUtils.cpp

```diff
@@ -24,151 +24,204 @@
+
 //FIXME: Consider to use variable to enable/disable cache state since using macro will not be able to close it if there're serious bugs.
 // #undef CC_ENABLE_GL_STATE_CACHE
 // #define CC_ENABLE_GL_STATE_CACHE 0
 
 #if CC_ENABLE_GL_STATE_CACHE
-
-namespace {
-    const int MAX_ACTIVE_TEXTURE = 16;
-    GLuint __currentArrayBufferId = -1;
-    GLuint __currentElementArrayBufferId = -1;
-    GLuint __activeTextureUnit = 0;
-    GLuint __currentBoundTexture[MAX_ACTIVE_TEXTURE] =  {
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1
-    };
+namespace
+{
+    GLint __currentVertexBuffer = -1;
+    GLint __currentIndexBuffer = -1;
+    GLint __currentVertexArray = -1;
+    
+    uint32_t __enabledVertexAttribArrayFlag = 0;
+    VertexAttributePointerInfo __enabledVertexAttribArrayInfo[MAX_ATTRIBUTE_UNIT];
 }
-
 #endif // CC_ENABLE_GL_STATE_CACHE
 
 //FIXME: need to consider invoking this after restarting game.
 void ccInvalidateStateCache()
 {
 #if CC_ENABLE_GL_STATE_CACHE
-    __currentArrayBufferId = -1;
-    __currentElementArrayBufferId = -1;
-    __activeTextureUnit = 0;
-    for (int i = 0; i < MAX_ACTIVE_TEXTURE; ++i)
-    {
-        __currentBoundTexture[i] = -1;
-    };
+    __currentVertexBuffer = -1;
+    __currentIndexBuffer = -1;
+    __currentVertexArray = -1;
+    
+    __enabledVertexAttribArrayFlag = 0;
+    for (int i = 0; i < MAX_ATTRIBUTE_UNIT; ++i)
+        __enabledVertexAttribArrayInfo[i] = VertexAttributePointerInfo();
 #endif
 }
 
+/****************************************************************************************
+ Buffer related
+ ***************************************************************************************/
+
 void ccBindBuffer(GLenum target, GLuint buffer)
 {
 #if CC_ENABLE_GL_STATE_CACHE
     if (target == GL_ARRAY_BUFFER)
     {
-        if (buffer == 0)
-            return;
-        if (__currentArrayBufferId != buffer)
+        if (__currentVertexBuffer != buffer)
         {
-            __currentArrayBufferId = buffer;
+            __currentVertexBuffer = buffer;
             glBindBuffer(target, buffer);
         }
-//        else
-//        {
-//            printf("glBindBuffer(GL_ARRAY_BUFFER, %u) isn't needed!\n", buffer);
-//        }
     }
     else if (target == GL_ELEMENT_ARRAY_BUFFER)
     {
-        if (buffer == 0)
-            return;
-        if (__currentElementArrayBufferId != buffer)
+        if (__currentIndexBuffer != buffer)
         {
-            __currentElementArrayBufferId = buffer;
+            __currentIndexBuffer = buffer;
             glBindBuffer(target, buffer);
         }
-//        else
-//        {
-//            printf("glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, %u) isn't needed!\n", buffer);
-//        }
     }
     else
     {
         glBindBuffer(target, buffer);
     }
+
+    CHECK_GL_ERROR_DEBUG();
```

Comment:
It's better to CHECK GL Error outside `ccBindBuffer` because we just wanna simulate `gl` functions which will not do this.

Reply:
Yep, agree.